cmd/cert-checker/main.go

```diff
@@ -0,0 +1,217 @@
+// Copyright 2015 ISRG.  All rights reserved
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package main
+
+import (
+	"crypto/x509"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"runtime"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/letsencrypt/boulder/Godeps/_workspace/src/github.com/cactus/go-statsd-client/statsd"
+	"github.com/letsencrypt/boulder/Godeps/_workspace/src/github.com/codegangsta/cli"
+	gorp "github.com/letsencrypt/boulder/Godeps/_workspace/src/gopkg.in/gorp.v1"
+	"github.com/letsencrypt/boulder/policy"
+
+	"github.com/letsencrypt/boulder/cmd"
+	"github.com/letsencrypt/boulder/core"
+	blog "github.com/letsencrypt/boulder/log"
+	"github.com/letsencrypt/boulder/sa"
+)
+
+const (
+	good = "valid"
+	bad  = "invalid"
+
+	filenameLayout = "20060102"
+
+	checkPeriod = time.Hour * 24 * 90
+)
+
+type report struct {
+	GoodCerts int64
+	BadCerts  int64
+	Entries   map[string]reportEntry
+}
+
+func (r *report) save(directory string) error {
+	filename := path.Join(directory, fmt.Sprintf(
+		"%s-%s-report.json",
+		time.Now().Add(-time.Hour*24*90).Format(filenameLayout),
+		time.Now().Format(filenameLayout),
+	))
+	content, err := json.Marshal(r)
+	if err != nil {
+		return err
+	}
+	return ioutil.WriteFile(filename, content, os.ModePerm)
+}
+
+type reportEntry struct {
+	Valid    bool     `json:"validity"`
+	Problems []string `json:"problem,omitempty"`
+}
+
+type certChecker struct {
+	pa           core.PolicyAuthority
+	dbMap        *gorp.DbMap
+	certs        chan core.Certificate
+	issuedReport report
+}
+
+func newChecker(dbMap *gorp.DbMap) certChecker {
+	c := certChecker{
+		pa:    policy.NewPolicyAuthorityImpl(),
+		dbMap: dbMap,
+	}
+	c.issuedReport.Entries = make(map[string]reportEntry)
+	return c
+}
+
+func (c *certChecker) getCerts() error {
+	var certs []core.Certificate
+	_, err := c.dbMap.Select(
+		&certs,
+		"SELECT * FROM certificates WHERE issued > :issued",
+		map[string]interface{}{"issued": time.Now().Add(-checkPeriod)},
+	)
+	if err != nil {
+		return err
+	}
+	c.certs = make(chan core.Certificate, len(certs))
+	for _, cert := range certs {
+		c.certs <- cert
+	}
+	// Close channel so range operations won't block when the channel empties out
+	close(c.certs)
```

Comment:
Hm? The WaitGroup is incremented/finished by the `processCerts` worker calls around L202, this call just collects all of the certificates for workers to process. The channel closed in order not to block `range` statements when it's empty, otherwise the workers will never break the `for ... range` and call `wg.Done()` and it will hang with all Goroutines sleeping waiting for something to come down the channel.

The SQL select above should probably be batched through with a `LIMIT` and add to this channel until it contains everything then `close` it though.